File: js/src/utils/reverseLookupBatch.ts
import { Connection, PublicKey } from "@solana/web3.js";
import { NameRegistryState } from "../state";

import { deserializeReverse } from "./deserializeReverse";
import {getReverseKeyFromDomainKey} from "./getReverseKeyFromDomainKey";

/**
 * This function can be used to perform a reverse look up
 * @param connection The Solana RPC connection
 * @param nameAccounts The public keys of the domains to look up
 * @returns The human readable domain names
 */
export async function reverseLookupBatch(
  connection: Connection,
  nameAccounts: PublicKey[],
): Promise<(string | undefined)[]> {
  let reverseLookupAccounts: PublicKey[] = [];
  for (let nameAccount of nameAccounts) {
    reverseLookupAccounts.push(getReverseKeyFromDomainKey(nameAccount));
  }

  let names = await NameRegistryState.retrieveBatch(
    connection,
    reverseLookupAccounts,
  );

  return names.map((name) => {
    if (name === undefined || name.data === undefined) {
      return undefined;
    }
    return deserializeReverse(name.data);
  });
}

File: python/utils/reverse_lookup_batch.py
from typing import List

from solana.rpc.async_api import AsyncClient
from solders.pubkey import Pubkey

from NameRegistryState import NameRegistryState
from utils.deserialize_reverse import deserialize_reverse
from utils.get_reverse_key_from_domain_key import get_reverse_key_from_domain_key


async def reverse_lookup_batch(
    connection: AsyncClient, name_accounts: List[Pubkey]
) -> List[str | None]:
    reverse_lookup_accounts: List[Pubkey] = []
    for name_account in name_accounts:
        reverse_lookup_accounts.append(get_reverse_key_from_domain_key(name_account))

    names = await NameRegistryState.retrieve_batch(connection, reverse_lookup_accounts)
    return [
        deserialize_reverse(name.data)
        if (name is not None and name.data is not None)
        else None
        for name in names
    ]
